### PR TITLE
Update to Calico 3.6

### PR DESCRIPTION
**Pull request by Cynerva**
Todo:
* [x] Update components to Calico 3.6
* [x] Support upgrades from Calico 2.x on amd64
* [x] Support upgrades from Calico 2.x on arm64

### CODE REVIEW

**Comment by Cynerva** [comment]
I've tested on both amd64 and arm64 and the results were good. This is ready for review.

File: build-calico-resource.sh [hunk 1]
```diff
@@ -50,29 +52,47 @@ for arch in ${arches}; do
...
 
   popd
   rm -rf resource-build-$arch
 done
+
+# calico-upgrade resource
+for arch in ${arches}; do
+  rm -rf resource-build-upgrade
+  mkdir resource-build-upgrade
+  pushd resource-build-upgrade
```
Comment (hyperbolic2346) [inline build-calico-resource.sh]: curious why pushd/popd vs cd/cd ..

File: reactive/calico.py [hunk 2]
```diff
@@ -21,19 +26,96 @@
...
+        message = 'Calico upgrade failed, see debug log'
+        status_set('blocked', message)
+        return
+    leader_set({'calico-v3-data-migration-needed': None})
+
+
+@when('leadership.is_leader')
+@when_not('leadership.set.calico-v3-data-migration-needed')
+def v3_data_ready():
+    leader_set({'calico-v3-data-ready': True})
```
Comment (hyperbolic2346) [inline reactive/calico.py]: We only ever seem to set this to true. Do we need this data?

File: reactive/calico.py [hunk 3]
```diff
@@ -21,19 +26,96 @@
...
+    try:
+        calico_upgrade.configure()
+        calico_upgrade.dry_run()
+        calico_upgrade.start()
+    except Exception:
+        log(traceback.format_exc())
+        message = 'Calico upgrade failed, see debug log'
+        status_set('blocked', message)
+        return
+    leader_set({'calico-v3-data-migration-needed': None})
```
Comment (hyperbolic2346) [inline reactive/calico.py]: So when the upgrade fails we set this to None. How would the user recover from this?

File: reactive/calico.py [hunk 4]
```diff
@@ -21,19 +26,96 @@
...
+            log('Failed to cleanup %s %s %s' % (kind, namespace, name))
+            return
+
+    leader_set({'calico-v3-npc-cleanup-needed': None})
+
+
+@when('leadership.is_leader', 'leadership.set.calico-v3-completion-needed',
+      'leadership.set.calico-v3-data-ready', 'calico.binaries.installed',
+      'calico.service.installed', 'calico.npc.deployed')
+@when_not('leadership.set.calico-v3-npc-cleanup-needed')
```
Comment (hyperbolic2346) [inline reactive/calico.py]: We don't need to worry about data migration being None here?

File: reactive/calico.py [hunk 5]
```diff
@@ -145,41 +228,52 @@ def install_calico_service():
...
     ''' Configure Calico IP pool. '''
     status_set('maintenance', 'Configuring Calico IP pool')
-    etcd = endpoint_from_flag('etcd.available')
-    env = os.environ.copy()
-    env['ETCD_ENDPOINTS'] = etcd.get_connection_string()
-    env['ETCD_KEY_FILE'] = ETCD_KEY_PATH
-    env['ETCD_CERT_FILE'] = ETCD_CERT_PATH
-    env['ETCD_CA_CERT_FILE'] = ETCD_CA_PATH
+
+    # remove unrecognized pools
```
Comment (hyperbolic2346) [inline reactive/calico.py]: Could the user have some pools they created that shouldn't be deleted?

File: config.yaml [hunk 6]
```diff
@@ -2,19 +2,19 @@ options:
...
   calico-policy-image:
     type: string
-    default: quay.io/calico/kube-controllers:v1.0.5
+    default: docker.io/calico/kube-controllers:v3.6.1
     description: |
       The image id to use for calico/kube-controllers.
   ipip:
-    type: boolean
-    default: false
+    type: string
```
Comment (hyperbolic2346) [inline config.yaml]: What's going to happen here on upgrade from a boolean?

File: build-calico-resource.sh (hunk 1 above)
Comment (Cynerva) [reply build-calico-resource.sh]: They were used elsewhere in the file. I used them for consistency.

I personally prefer this anyway:
```
(cd
  do stuff
)
```

File: reactive/calico.py (hunk 2 above)
Comment (Cynerva) [reply reactive/calico.py]: Yes. It is `None` by default. It is used here to determine whether or not an upgrade is needed:

https://github.com/charmed-kubernetes/layer-calico/blob/c4ec07666ab118d84f1c99986b4024815a4f6cc3/reactive/calico.py#L45

File: reactive/calico.py (hunk 4 above)
Comment (Cynerva) [reply reactive/calico.py]: `leadership.set.calico-v3-data-ready` implies `not leadership.set.calico-v3-data-migration-needed`. We don't need both, but I can add the latter if it makes sense to you.

File: reactive/calico.py (hunk 5 above)
Comment (Cynerva) [reply reactive/calico.py]: We don't support user-created pools, currently, but we will soon with the BGP work.

My original design for Calico BGP called for an `enable-default-pool` config. It made sense to me to change that to `manage-pools` and have it cover the deletion of unrecognized pools, too. That's what I'm going for.

It is possible to have this code only delete pools with conflicting CIDRs. The logic for that isn't too bad actually. But I figure with the `enable-default-pools` / `manage-pools` config coming up soon, it's not really needed.

File: config.yaml (hunk 6 above)
Comment (Cynerva) [reply config.yaml]: If it's set to the default, it upgrades to the new default value just fine.

No idea what happens if you upgrade when the value is not the default. I'm guessing it shouldn't be catastrophic?

File: reactive/calico.py (hunk 2 above)
Comment (hyperbolic2346) [reply reactive/calico.py]: I guess I'm getting tripped up on the naming. It isn't obvious that this indicated that an upgrade is needed or that it implies we don't need a data migration. Unfortunately, I'm drawing a blank for helpful suggestions.

File: reactive/calico.py (hunk 5 above)
Comment (hyperbolic2346) [reply reactive/calico.py]: As long as we have it worked out how a user can create and manage pools, I'm ok with this. I just wanted to ensure we thought about the upgrade scenario.

File: config.yaml (hunk 6 above)
Comment (hyperbolic2346) [reply config.yaml]: Would that be an easy thing for you to test at this point?

File: reactive/calico.py (hunk 2 above)
Comment (Cynerva) [reply reactive/calico.py]: Some alternative names:

calico-v3-data-available
calico-v3-components-runnable

Another option is to remove `calico-v3-data-ready`, and invert the truthiness of `calico-v3-data-migration-needed` (so it becomes something like `calico-v3-data-migration-complete`) so followers can wait for that instead. What do you think?

File: config.yaml (hunk 6 above)
Comment (Cynerva) [reply config.yaml]: Gave it a quick test. It just reverts to the new default. That can go in a release note.

File: reactive/calico.py (hunk 4 above)
Comment (hyperbolic2346) [reply reactive/calico.py]: No, I think we're good. I was just confused by the ordering. The migration happens before the data is ready, which isn't how I read it the first time through. I think I've got it now and this is ok.

**Pull request merged**